compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp

```
@@ -0,0 +1,227 @@
+  : _input_data_path(input_data_path), _qerror_ratio(qerror_ratio), _qerror(0.f),
+    _quantizer(std::make_unique<Quantizer>("uint8", "uint8"))
+{
+}
+
+BisectionSolver::Options *BisectionSolver::options(void)
+{
+  if (_options == nullptr)
+  {
+    _options = std::make_unique<BisectionOptionsImpl>();
+  }
+
+  return _options.get();
+}
+
+float BisectionSolver::evaluate(const DatasetEvaluator &evaluator, const std::string &flt_path,
+                                const std::string &def_quant, LayerParams &layers)
+{
+  auto model = read_module(flt_path);
+  // get fake quantized model for evaluation
+  if (!_quantizer->fake_quantize(model.get(), def_quant, layers))
+  {
+    throw std::runtime_error("Failed to produce fake-quantized model.");
+  }
+
+  return evaluator.evaluate(model.get());
+}
+
+std::unique_ptr<luci::Module> BisectionSolver::run(const std::string &module_path)
+{
+  LOGGER(l);
+
+  auto module = read_module(module_path);
+
+  float min_depth = 0.f;
+  float max_depth = 0.f;
+  NodeDepthType nodes_depth;
+  if (compute_depth(module.get(), nodes_depth, min_depth, max_depth) !=
+      ParameterizerResult::SUCCESS)
+  {
+    std::cerr << "Invalid graph for bisectioning" << std::endl;
+    return nullptr;
+  }
+
+  std::unique_ptr<MAEMetric> metric = std::make_unique<MAEMetric>();
+  DatasetEvaluator evaluator(module.get(), _input_data_path, *metric.get());
+
+  // const auto &ref_output = compute_outputs(module.get(), _input_data_path);
+  LayerParams layer_params;
+  float int16_qerror = evaluate(evaluator, module_path, "int16", layer_params);
+  VERBOSE(l, 0) << "int16_quantization_error " << int16_qerror << std::endl;
+
+  float int8_qerror = evaluate(evaluator, module_path, "uint8", layer_params);
+  VERBOSE(l, 0) << "int8_quantization_error " << int8_qerror << std::endl;
+
+  _qerror = int16_qerror + _qerror_ratio * std::fabs(int8_qerror - int16_qerror);
```

Please throw an exception if int8_qerror < int16_qerror.

std::runtime_error("Q8 model's qerror is less than Q16 model's qerror.");

---

ok